WallsAndHoles/xmltool.cpp

```diff
@@ -0,0 +1,283 @@
+            QMessageBox::Ok);
+            return nullptr;
+    }
+    //close reader and flush file
+    xmlReader.clear();
+    file.close();
+    return templateSet;
+}
+
+int XMLTool::saveTileMap(SharedTileMap tileMap, bool saveTemplates){
+    QVector<SharedTileTemplateSet> templateSets = tileMap->dependencies();
+    const Array2D<QSharedPointer<Tile>> tiles = tileMap->cTiles();
+    QString tileMapPath = tileMap->savePath();
+    QFile file(tileMapPath);
+    if (!file.open(QIODevice::WriteOnly | QIODevice::Truncate))
+        return -1;
+    QTextStream out(&file);
+    QDomDocument doc;
+    QDomElement element;
+    QDomAttr attr;
+    QDomProcessingInstruction instruction = doc.createProcessingInstruction( "xml", "version = \'1.0\' encoding=\'UTF-8\'" );
+    doc.appendChild(instruction);
+
+    QDomElement root = doc.createElement( "TileMap" );
+    attr = doc.createAttribute("width");
+    attr.setValue(QString::number(tiles.size().width()));
+    root.setAttributeNode(attr);
+    attr = doc.createAttribute("height");
+    attr.setValue(QString::number(tiles.size().height()));
+    root.setAttributeNode(attr);
+    doc.appendChild(root);
+    for(int i=0; i<templateSets.size(); i++){
+        element = doc.createElement("TileTemplateSet");
+        attr = doc.createAttribute("id");
+        attr.setValue(QString::number(i));
+        element.setAttributeNode(attr);
+        attr = doc.createAttribute("SavePath");
+        attr.setValue(templateSets[i]->savePath());
+        element.setAttributeNode(attr);
+        root.appendChild(element);
+        if(saveTemplates){ XMLTool::saveTileTemplateSet(templateSets[i]); }
+    }
+    for(int i=0; i<tiles.size().height(); i++){
+        for(int j=0; j<tiles.size().width(); j++){
+            QSharedPointer<Tile> tile = tiles(i,j);
+            if(tile->hasTileTemplate()){
+                element = doc.createElement("Tile");
+                attr = doc.createAttribute("x");
+                attr.setValue(QString::number(j));
+                element.setAttributeNode(attr);
+                attr = doc.createAttribute("y");
+                attr.setValue(QString::number(i));
+                element.setAttributeNode(attr);
+                attr = doc.createAttribute("RelativeThickness");
+                attr.setValue(QString::number(tile->relativeThickness()));
+                element.setAttributeNode(attr);
+                attr = doc.createAttribute("RelativeHeight");
+                attr.setValue(QString::number(tile->relativeHeight()));
+                element.setAttributeNode(attr);
+                attr = doc.createAttribute("RelativePosition");
+                attr.setValue(QString("%1,%2").arg(
+                    QString::number(tile->relativePosition()[0]),
+                    QString::number(tile->relativePosition()[1])));
+                element.setAttributeNode(attr);
+                attr = doc.createAttribute("TemplateID");
+                int id=-1; int count=0;
```

Comment:
It would work a bit nicer for loading it instead of just one id, it stored one for the id of tileTemplateSet, and another for the TileTemplate within that set.

Reply:
I am imitating tiled map editor's choice. It use single id to index all tile, so it can pack entire tile-template relation into a matrix. (Honestly, I believe use id instead of pointer is a better way to model tilemap, it treat tileTemplate like sort of "color" and then tilemap can be treated like sort of image, then many image processing algorithm can be applied immediately)